### PR TITLE
IGNITE-20181: KV/Binary view public API should only throw public exceptions for the end user.

### PR DESCRIPTION
**Pull request by lowka**
Wraps marshaller exceptions in lang.MarshallerException and fixes MarshallerException exception error code/ message propagation by adding the required constructor.

https://issues.apache.org/jira/browse/IGNITE-20181

----

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/api/src/main/java/org/apache/ignite/lang/MarshallerException.java [hunk 1]
```diff
@@ -28,6 +32,18 @@ public class MarshallerException extends IgniteException {
...
+
+    /**
+     * Creates an exception with the given trace ID, error code, detailed message, and cause.
+     *
+     * @param traceId Unique identifier of the exception.
+     * @param code Full error code.
+     * @param message Detailed message.
+     * @param cause Optional nested exception (can be {@code null}).
+     */
+    public MarshallerException(UUID traceId, int code, String message, @Nullable Throwable cause) {
```
Comment (lowka) [inline modules/api/src/main/java/org/apache/ignite/lang/MarshallerException.java:46]: This constructor is need to preserve information provided with exception created via `MarshallerException(Throwable cause)`.

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java [hunk 2]
```diff
@@ -68,15 +71,18 @@ protected <T> T sync(CompletableFuture<T> fut) {
...
      * @return Public exception.
      */
-    protected IgniteException convertException(Throwable th) {
-        if (th instanceof IgniteException) {
-            return (IgniteException) th;
+    protected IgniteException convertException(Throwable t) {
+        if (t instanceof IgniteException) {
+            return (IgniteException) t;
+        } else if (t instanceof IgniteInternalCheckedException || t instanceof IgniteInternalException) {
+            // Map internal errors to IgniteException with code internal_err.
```
Comment (ygerzhedovich) [inline modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:81]: why you have two `else` block with `instanceof` instead just return `new IgniteException(Common.INTERNAL_ERR, t)`?

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java [hunk 3]
```diff
@@ -68,15 +69,15 @@ protected <T> T sync(CompletableFuture<T> fut) {
...
      * Converts an internal exception to a public one.
      *
-     * @param th Internal exception.
+     * @param t Internal exception.
      * @return Public exception.
      */
-    protected IgniteException convertException(Throwable th) {
-        if (th instanceof IgniteException) {
-            return (IgniteException) th;
+    protected IgniteException convertException(Throwable t) {
```
Comment (ygerzhedovich) [inline modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:75]: why you didn't use org.apache.ignite.lang.IgniteExceptionMapperUtil#mapToPublicException ?

Comment (lowka) [reply modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java]: Fixed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java [hunk 4]
```diff
@@ -31,7 +32,7 @@ abstract class AbstractTableView {
     protected final InternalTable tbl;
 
     /** Schema registry. */
-    protected final SchemaRegistry schemaReg;
+    protected final TableViewRowConverter rowConverter;
```
Comment (xtern) [inline modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:35]: Please change the description as well.

File: modules/api/src/main/java/org/apache/ignite/lang/UnexpectedNullValueException.java [hunk 5]
```diff
@@ -30,7 +31,7 @@ public class UnexpectedNullValueException extends IgniteException {
      * @param msg Message.
      */
     public UnexpectedNullValueException(String msg) {
-        super("Got unexpected null value: " + msg);
+        super(Common.INTERNAL_ERR, "Got unexpected null value: " + msg);
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/api/src/main/java/org/apache/ignite/lang/UnexpectedNullValueException.java]: Fixed.

File: modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java (hunk 4 above)
Comment (lowka) [reply modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java:35]: Fixed.

**Comment by lowka** [comment]
Removed exception handling wrapper of `InternalTable` and wrapped read/modify calls of `InternalTable ` with `covertToPublicFuture` .

**Review by xtern — APPROVED** [review]

**Review by ygerzhedovich — APPROVED** [review]

**Pull request merged**